File: platform/test/connection/openai-embeddings/oss-undici/test_case.mjs
#!/usr/bin/env node

/**
 * Connection test for OpenAI Embeddings API via raw open-source undici library.
 *
 * Tests proxy, verify_ssl, ca_bundle, and client cert configuration
 * by resolving config from AppYamlConfig and making a health-check request
 * using undici's low-level request() API with ProxyAgent / Agent dispatchers.
 */

import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { request, ProxyAgent, Agent } from 'undici';
import fs from 'node:fs';

const args = process.argv.slice(2);
const readArg = (name, fallback) => {
  const idx = args.indexOf(name);
  return idx >= 0 ? args[idx + 1] : fallback;
};

const appEnv = readArg('--app-env', process.env.APP_ENV || 'dev');
const configDir = readArg('--config-dir', process.env.CONFIG_DIR);
if (!configDir) {
  console.error('Missing --config-dir');
  process.exit(2);
}

const platformRoot = path.resolve(configDir, '..', '..');
const importFile = async (absPath) => import(pathToFileURL(absPath).href);

const appYamlLoadMod = await importFile(path.join(platformRoot, 'polyglot/app_yaml_load/mjs/dist/index.js'));

const { config: appConfig } = await appYamlLoadMod.loadAppYamlConfig({ configDir, appEnv });
const providerConfig = appConfig.getNested(['providers', 'openai_embeddings'], {}) || {};

// ---------------------------------------------------------------------------
// Connection config resolution
// ---------------------------------------------------------------------------

const connCfg = {
  proxyUrl: providerConfig.proxy_url || process.env.HTTPS_PROXY || process.env.HTTP_PROXY || null,
  verifySsl: providerConfig.verify_ssl ?? true,
  caBundlePath: providerConfig.ca_bundle || process.env.SSL_CERT_FILE || process.env.NODE_EXTRA_CA_CERTS || null,
  clientCert: providerConfig.client_cert || null,
  clientKey: providerConfig.client_key || null,
};

// ---------------------------------------------------------------------------
// Connection test
// ---------------------------------------------------------------------------

const resolvedApiKey = providerConfig.api_key || process.env.OPENAI_EMBEDDINGS_API_KEY || process.env.OPENAI_API_KEY;

// Treat unresolved template placeholders as missing
const apiKey = resolvedApiKey && !resolvedApiKey.startsWith('{{') ? resolvedApiKey : null;

let connectionTest;

if (!apiKey) {
  connectionTest = {
    skipped: true,
    reason: 'no OPENAI_EMBEDDINGS_API_KEY, OPENAI_API_KEY or api_key in config',
  };
} else {
  try {
    const baseUrl = providerConfig.base_url || 'https://api.openai.com/v1';

    // Build TLS connect options
    const connectOptions = {};
    if (connCfg.caBundlePath) {
      connectOptions.ca = fs.readFileSync(connCfg.caBundlePath);
    }
    if (connCfg.clientCert) {
      connectOptions.cert = fs.readFileSync(connCfg.clientCert);
    }
    if (connCfg.clientKey) {
      connectOptions.key = fs.readFileSync(connCfg.clientKey);
    }
    // CodeQL: js/disabling-certificate-validation — only disable cert validation
    // when the connection config explicitly opts out AND we are in a test context
    if (!connCfg.verifySsl && process.env.NODE_ENV === 'test') {
      connectOptions.rejectUnauthorized = false;
    }

    // Create dispatcher (ProxyAgent or Agent)
    let dispatcher;
    if (connCfg.proxyUrl) {
      dispatcher = new ProxyAgent({
        uri: connCfg.proxyUrl,
        requestTls: connectOptions,
      });
    } else if (Object.keys(connectOptions).length > 0) {
      dispatcher = new Agent({ connect: connectOptions });
    }

    const url = `${baseUrl}/embeddings`;
    const start = performance.now();

    const { statusCode, headers, body } = await request(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': 'openai-embeddings-connection-test/1.0',
      },
      body: JSON.stringify({ model: 'text-embedding-3-small', input: 'Hello, world!' }),
      dispatcher,
    });

    const data = await body.json();
    const latency = Math.round(performance.now() - start);

    connectionTest = {
      skipped: false,
      connected: statusCode >= 200 && statusCode < 300,
      endpoint: '/embeddings',
      status_code: statusCode,
      hasEmbedding: data.data?.[0]?.embedding?.length > 0,
      latency_ms: latency,
      connection: {
        proxy: connCfg.proxyUrl || '(none)',
        verify_ssl: connCfg.verifySsl,
        ca_bundle: connCfg.caBundlePath || '(default)',
        has_client_cert: Boolean(connCfg.clientCert),
      },
    };
  } catch (err) {
    connectionTest = {
      skipped: false,
      connected: false,
      error: { name: err.name, message: err.message },
    };
  }
}

const summary = {
  service: 'openai-embeddings-oss-undici',
  mode: 'connection-test',
  mjs_package: 'undici:request',
  app_yaml_initialized: true,
  connection_config: connCfg,
  connection_test: connectionTest,
  overall: connectionTest.connected ? 'PASS' : connectionTest.skipped ? 'SKIP' : 'FAIL',
};

// Redact sensitive fields before logging (CodeQL js/clear-text-logging)
const redactedSummary = { ...summary, connection_config: { ...summary.connection_config, proxyUrl: summary.connection_config.proxyUrl ? '*******' : null, clientKey: summary.connection_config.clientKey ? '*******' : null } };
console.log(JSON.stringify(redactedSummary, null, 2));

File: platform/test/connection/openai-embeddings/native-py/test_case.py
#!/usr/bin/env python3
"""
Connection test for OpenAI Embeddings API via Python stdlib (urllib.request).

Tests proxy, verify_ssl, ca_bundle, and client cert configuration
by resolving config from AppYamlConfig and making a health-check request
using only native Python modules — no third-party HTTP libraries.
"""

from __future__ import annotations

import argparse
import asyncio
import json
import os
import re
import ssl
import sys
import time
import urllib.request
from pathlib import Path

_HERE = Path(__file__).resolve().parent
_PLATFORM = _HERE
while _PLATFORM.name != "platform" and _PLATFORM != _PLATFORM.parent:
    _PLATFORM = _PLATFORM.parent


def _inject_paths():
    """Add polyglot/packages_py source dirs to sys.path from platform pyproject.toml."""
    toml = _PLATFORM / "pyproject.toml"
    if not toml.exists():
        return
    for line in toml.read_text().splitlines():
        m = re.search(r'path\s*=\s*"([^"]+)"', line)
        if not m:
            continue
        pkg_path = (_PLATFORM / m.group(1)).resolve()
        src = pkg_path / "src"
        sys.path.insert(0, str(src if src.is_dir() else pkg_path))


_inject_paths()

from app_yaml_load import load_app_yaml_config


# ---------------------------------------------------------------------------
# Connection config resolution
# ---------------------------------------------------------------------------

def _resolve_connection_config(provider_cfg: dict) -> dict:
    """Resolve proxy, SSL, and cert settings from YAML + env."""
    return {
        "proxy_url": (
            provider_cfg.get("proxy_url")
            or os.environ.get("HTTPS_PROXY")
            or os.environ.get("HTTP_PROXY")
        ),
        "verify_ssl": provider_cfg.get("verify_ssl", True),
        "ca_bundle": (
            provider_cfg.get("ca_bundle")
            or os.environ.get("SSL_CERT_FILE")
            or os.environ.get("REQUESTS_CA_BUNDLE")
        ),
        "client_cert": provider_cfg.get("client_cert"),
        "client_key": provider_cfg.get("client_key"),
    }


def _resolve_api_key(provider_cfg: dict) -> str | None:
    """Resolve API key from config (with template guard) or env fallbacks."""
    cfg_key = provider_cfg.get("api_key") or ""
    if cfg_key and not cfg_key.startswith("{{"):
        return cfg_key
    return (
        os.environ.get("OPENAI_EMBEDDINGS_API_KEY")
        or os.environ.get("OPENAI_API_KEY")
    )


def _build_ssl_context(conn_cfg: dict) -> ssl.SSLContext:
    """Build an ssl.SSLContext from connection config."""
    if not conn_cfg["verify_ssl"]:
        ctx = ssl.create_default_context()
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE
    elif conn_cfg["ca_bundle"]:
        ctx = ssl.create_default_context(cafile=conn_cfg["ca_bundle"])
    else:
        ctx = ssl.create_default_context()

    if conn_cfg["client_cert"]:
        ctx.load_cert_chain(
            certfile=conn_cfg["client_cert"],
            keyfile=conn_cfg["client_key"],
        )
    return ctx


def _test_connection(provider_cfg: dict, conn_cfg: dict) -> dict:
    """Test OpenAI Embeddings API connection using urllib.request."""
    api_key = _resolve_api_key(provider_cfg)
    if not api_key:
        return {
            "skipped": True,
            "reason": "no OPENAI_EMBEDDINGS_API_KEY, OPENAI_API_KEY, or api_key in config",
        }

    base_url = provider_cfg.get("base_url") or "https://api.openai.com/v1"
    url = f"{base_url}/embeddings"

    ssl_ctx = _build_ssl_context(conn_cfg)

    # Proxy handler
    proxy_url = conn_cfg["proxy_url"]
    if proxy_url:
        proxy_handler = urllib.request.ProxyHandler({"https": proxy_url, "http": proxy_url})
        opener = urllib.request.build_opener(
            proxy_handler,
            urllib.request.HTTPSHandler(context=ssl_ctx),
        )
    else:
        opener = urllib.request.build_opener(
            urllib.request.HTTPSHandler(context=ssl_ctx),
        )

    payload = json.dumps({"model": "text-embedding-3-small", "input": "Hello, world!"}).encode("utf-8")

    req = urllib.request.Request(
        url,
        data=payload,
        method="POST",
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
            "User-Agent": "openai-embeddings-connection-test/1.0",
        },
    )

    try:
        start = time.monotonic()
        with opener.open(req, timeout=15) as resp:
            status_code = resp.status
            raw = resp.read()
        latency = round((time.monotonic() - start) * 1000)

        data = json.loads(raw)
        embedding = data.get("data", [{}])[0].get("embedding", [])

        return {
            "skipped": False,
            "connected": status_code < 400,
            "status_code": status_code,
            "endpoint": "POST /embeddings",
            "has_embedding": len(embedding) > 0,
            "embedding_length": len(embedding),
            "latency_ms": latency,
            "connection": {
                "base_url": base_url,
                "proxy": proxy_url or "(none)",
                "verify_ssl": conn_cfg["verify_ssl"],
                "ca_bundle": conn_cfg["ca_bundle"] or "(default)",
                "has_client_cert": bool(conn_cfg["client_cert"]),
            },
        }
    except urllib.error.HTTPError as exc:
        return {
            "skipped": False,
            "connected": False,
            "status_code": exc.code,
            "error": {"name": "HTTPError", "message": str(exc)},
        }
    except Exception as exc:
        return {
            "skipped": False,
            "connected": False,
            "error": {"name": type(exc).__name__, "message": str(exc)},
        }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-env", default="dev")
    parser.add_argument("--config-dir", required=True)
    args = parser.parse_args()

    app_cfg = load_app_yaml_config(config_dir=args.config_dir, app_env=args.app_env).config
    provider_cfg = app_cfg.get_nested("providers", "openai_embeddings", default={}) or {}
    conn_cfg = _resolve_connection_config(provider_cfg)

    connection_test = _test_connection(provider_cfg, conn_cfg)

    summary = {
        "service": "openai-embeddings-native-py",
        "mode": "connection-test",
        "python_module": "urllib.request",
        "app_yaml_initialized": True,
        "connection_config": conn_cfg,
        "connection_test": connection_test,
        "overall": "PASS" if connection_test.get("connected") else (
            "SKIP" if connection_test.get("skipped") else "FAIL"
        ),
    }
    print(json.dumps(summary, indent=2, default=str))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
